google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java

```diff
@@ -260,6 +263,79 @@ public void downloadTo(Path path) {
     downloadTo(path, new BlobSourceOption[0]);
   }
 
+  /**
+   * Uploads the given file path to this blob using specified blob write options.
+   *
+   * @param path file to be uploaded
+   * @param options blob write options
+   * @return updated blob
+   * @throws IOException on I/O error
+   * @throws StorageException on failure
+   */
+  public Blob uploadFrom(Path path, BlobWriteOption... options) throws IOException {
+    if (Files.isDirectory(path)) {
+      throw new StorageException(0, path + ": Is a directory");
+    }
+    try (InputStream input = Files.newInputStream(path)) {
+      return uploadFrom(input, options);
+    }
+  }
+
+  /**
+   * Uploads the given content to this blob using specified blob write options.
+   *
+   * @param input content to be uploaded
+   * @param options blob write options
+   * @return updated blob
+   * @throws IOException on I/O error
+   * @throws StorageException on failure
+   */
+  public Blob uploadFrom(InputStream input, BlobWriteOption... options) throws IOException {
+    try (WriteChannel writer = storage.writer(this, options)) {
+      upload(input, writer);
+    }
+    BlobId blobId = getBlobId();
+    return storage.get(BlobId.of(blobId.getBucket(), blobId.getName()));
+  }
+
+  static void upload(InputStream input, WriteChannel writer) throws IOException {
+    upload(input, writer, DEFAULT_CHUNK_SIZE);
+  }
+
+  /**
+   * Uploads the given content to the storage using specified write channel and the given
+   * buffer size. The default buffer size is 15 MiB. Larger buffer sizes may improve the upload
+   * performance but require more memory. It could cause OutOfMemoryError or add significant garbage
+   * collection overhead. Buffer sizes which are less than 256 KiB are not allowed, they will be
+   * treated as 256 KiB.
+   *
+   * <p>This method does not close either InputStream or WriterChannel</p>
```

Comment:
These details can be moved to the previous method.

Reply:
The previous one is intentionally made package private, it will be invoked from the StorageImpl class.

Reply:
I'm confused about the difference between upload and uploadFrom, so probably users will be too. Stocking with the public methods, I don't see any particular difference between the names `upload` and `uploadFrom`. I suggest making both `upload`.

Also, why does one return a blob and one doesn't. Both should or neither. 

Since I am confused, you might have good reasons for this distinction, but in that case let's make sure the docs and the method names make these reasons obvious.

Reply:
Returning a Blob instance after upload has completed requires an extra GET request which is not always needed. So, `upload` works faster than `uploadFrom`, especially on small files.
I'm going to document this difference in the new method: `Storage.upload(BlobInfo, Path)`

`uploadFrom` is named after `downloadTo` (as requested in the issue).
`upload` is static and its name is different. Would it be more clear to use `writeContnet` instead?

```
/**
  * Writes the given content to the storage using specified write channel and the given
  * buffer size.
  */
public static void Blob.writeContent(InputStream content, WriterChannel writer, int bufferSize)
```

Reply:
I'm not sure.The distinction between the names uploadFrom and upload seems very hazy. Honestly, if I had to guess I'd expect uploadFrom was static rather than the other way around.

Do we actually need both of these? Can we delete the one that makes an extra GET request?

Consider this case:

1. First request to upload a blob succeeds.
2. GET request fails.
3. Exception thrown

It won't be obvious to the user that the content was successfully upload to the server.

Reply:
Yes, this is a good example, quite rare but still valid.
To inform user of successful upload I re-throw an exception with the attendant message.

> Do we actually need both of these? Can we delete the one that makes an extra GET request?

Yes, I believe we need both. If  `blob.uploadFrom()` returns `void`, then `blob` instance will become useless right after upload.  As  #149 demonstrates, it's not obvious how to get the updated blob object, so returning the updated updated would be helpful.